### DISCUSSION TITLE
Architecture of shadow 2.x

### DISCUSSION THREAD

**Discussion by lorenzo9uerra** [shadow/shadow#1635] [Q&A] [answered]
Hi!
I understand this is still a pre-release and there could be breaking changes, but could someone explain to me the latest architecture and in what differs from shadow 1.x? It doesn't need to be in too much detail.

Any help would be really appreciated,
Thank you

**Comment by sporksmith** [comment, marked as answer]
Hi Inwor,
The main architectural difference is that each process and thread in the simulation is now represented as a real process or thread in the host Linux operating system. This removes a lot of complexity, makes it easier to use introspection tools like debuggers, and allows us to run unmodified binaries.

The [Show and tell](https://github.com/shadow/shadow/discussions/categories/show-and-tell) posts talk about the work we've been doing. I think the first one where we talk about the new architecture is here: https://lists.torproject.org/pipermail/tor-project/2020-March/002785.html

**Reply by lorenzo9uerra** [reply]
I read the show and tell posts and about the difference between plug-in and managed process, but I wanted to be sure that I wasn't missing any other major change.
Thank you

**Reply by bergerch** [reply]
Hi, thank you. I also find this question very interesting, in particular how Shadow handles the interposition between the real process and kernel. Also if there is a good write up that explains the basics of shadows internal workings (scheduling events, simulating the network, etc.; I think this remains similiar to the old version?) it would be great if you could link it here.

**Reply by sporksmith** [reply]
@bergerch by default, we use LD_PRELOAD to intercept at the C API level, and seccomp to intercept any syscalls that aren't caught by the first mechanism. There's also an option to use ptrace instead of seccomp, but ptrace is likely to go away in the 2.0 release since it's slower and has some other limitations.

We'll definitely have an updated writeup for the 2.0 release.
